crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -0,0 +1,453 @@
+use crate::{
+    traits::{BlockSource, ReceiptProvider},
+    BlockHashProvider, BlockNumProvider, BlockProvider, EvmEnvProvider, HeaderProvider,
+    ProviderError, StageCheckpointProvider, TransactionsProvider, WithdrawalsProvider,
+};
+use reth_db::{
+    cursor::DbCursorRO,
+    database::DatabaseGAT,
+    models::StoredBlockBodyIndices,
+    tables,
+    transaction::{DbTx, DbTxMut},
+};
+use reth_interfaces::Result;
+use reth_primitives::{
+    stage::{StageCheckpoint, StageId},
+    Block, BlockHash, BlockHashOrNumber, BlockNumber, ChainInfo, ChainSpec, Head, Header, Receipt,
+    SealedBlock, SealedHeader, TransactionMeta, TransactionSigned, TxHash, TxNumber, Withdrawal,
+    H256, U256,
+};
+use reth_revm_primitives::{
+    config::revm_spec,
+    env::{fill_block_env, fill_cfg_and_block_env, fill_cfg_env},
+    primitives::{BlockEnv, CfgEnv, SpecId},
+};
+use std::{ops::RangeBounds, sync::Arc};
+
+pub(crate) type ProviderRO<'this, DB> = Provider<'this, <DB as DatabaseGAT<'this>>::TX>;
+
+pub(crate) type ProviderRW<'this, DB> = Provider<'this, <DB as DatabaseGAT<'this>>::TXMut>;
+
+/// Provider struct which allows to interface with the database using different types of providers.
+/// Wrapper around [`DbTx`] and [`DbTxMut`]. Example: [`HeaderProvider`] [`BlockHashProvider`]
+pub struct Provider<'this, TX>
```

Comment:
this provider strictly operates on the DB, I suggest we name it `DatabaseProvider` to make this clear.